crates/ruff_python_parser/src/parser/statement.rs

```diff
@@ -413,52 +431,160 @@ impl<'src> Parser<'src> {
+        let mut leading_dots = if self.eat(TokenKind::Ellipsis) { 3 } else { 0 };
         let mut progress = ParserProgress::default();
 
         while self.at_ts(DOT_ELLIPSIS_SET) {
             progress.assert_progressing(self);
 
             if self.eat(TokenKind::Dot) {
-                level += 1;
-            }
-
-            if self.eat(TokenKind::Ellipsis) {
-                level += 3;
+                leading_dots += 1;
+            } else if self.eat(TokenKind::Ellipsis) {
+                leading_dots += 3;
             }
         }
 
         if self.at(TokenKind::Name) {
             module = Some(self.parse_dotted_name());
         };
 
-        if level == 0 && module.is_none() {
-            let range = self.current_token_range();
+        if leading_dots == 0 && module.is_none() {
+            // test_err from_import_missing_module
+            // from
+            // from import x
             self.add_error(
-                ParseErrorType::OtherError("missing module name".to_string()),
-                range,
+                ParseErrorType::OtherError("Expected a module name".to_string()),
+                self.current_token_range(),
             );
         }
 
+        // test from_import_no_space
+        // from.import x
+        // from...import x
         self.expect(TokenKind::Import);
 
-        let parenthesized = self.eat(TokenKind::Lpar);
+        let names_start = self.node_start();
+        let mut names = vec![];
+        let mut seen_star_import = false;
 
-        let names = self.parse_comma_separated_list_into_vec(
-            RecoveryContextKind::ImportFromAsNames,
-            Parser::parse_alias,
+        let parenthesized = Parenthesized::from(self.eat(TokenKind::Lpar));
+
+        // test_err from_import_unparenthesized_trailing_comma
+        // from a import b,
+        // from a import b as c,
+        // from a import b, c,
+        self.parse_comma_separated_list(
+            RecoveryContextKind::ImportFromAsNames(parenthesized),
+            |parser| {
+                let alias = parser.parse_alias();
+                seen_star_import |= alias.name.id == "*";
+                names.push(alias);
+            },
         );
 
-        // TODO(dhruvmanila): Error when `*` is mixed with other names.
+        if names.is_empty() {
+            // test_err from_import_empty_names
+            // from x import
+            // from x import ()
+            // from x import ,,
+            self.add_error(ParseErrorType::EmptyImportNames, self.current_token_range());
+        }
+
+        if seen_star_import && names.len() > 1 {
+            // test_err from_import_star_with_other_names
+            // from x import *, a
+            // from x import a, *, b
+            // from x import *, a as b
+            self.add_error(
+                ParseErrorType::OtherError("Star import must be the only import".to_string()),
+                self.node_range(names_start),
+            );
+        }
 
```

Comment:
What about `from x import *, *, a`?

Reply:
It does report an error. Is the worry that there are 2 star imports in your test case? I'll add it.